python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -29,8 +36,65 @@ class TFRecordDatasource(FileBasedDatasource):
 
     _FILE_EXTENSION = "tfrecords"
 
+    def __init__(self, filesystem: Optional["pyarrow.fs.FileSystem"]):
+        self._fs = filesystem
```

Comment:
Note: this `filesystem` assignment and the `_get_full_file_path` _should_ be unnecessary, as `_FileBasedDatasourceReader` is [already handling it here](https://github.com/ray-project/ray/blob/master/python/ray/data/datasource/file_based_datasource.py#L290). However, there is currently likely a bug related to this (filed an [issue](https://github.com/ray-project/ray/issues/33777) a couple weeks ago already). Once this issue is resolved, we should be able to remove this workaround.

To easily come back to this in the future, could we add a note here to track the workaround origin?
```suggestion
# TODO: remove filesystem assignment and _get_full_file_path() after https://github.com/ray-project/ray/issues/33777 is resolved
        self._fs = filesystem
```

Reply:
`_get_full_file_path ` is needed because the path provided in the `_read_stream ` is relative to the filesystem. 

Our current implementation of `_get_full_file_path` is not complete, and only works with local and gcs file systems